frame/pallet-xvm/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+//! # XVM pallet
+//!
+//! ## Overview
+//!
+//!
+//! ## Interface
+//!
+//! ### Dispatchable Function
+//!
+//!
+//! ### Other
+//!
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+use sp_runtime::{traits::Member, RuntimeDebug};
+use sp_std::prelude::*;
+
+pub mod pallet;
+pub use pallet::pallet::*;
+
+/// EVM call adapter instance.
+#[cfg(feature = "evm")]
+pub mod evm;
+
+/// Wasm call adapter instance.
+#[cfg(feature = "wasm")]
+pub mod wasm;
+
+/// XVM context consist of unique ID and optional execution arguments.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, scale_info::TypeInfo)]
+pub struct XvmContext<VmId> {
+    /// VM identifier.
+    pub id: VmId,
+    /// Encoded VM execution environment.
+    pub env: Option<Vec<u8>>,
```

Comment:
We've had discussion about this in slack but I still don't see why this is Vec<u8>.

It is true that different VMs have different view of e.g. gas (EVM vs wasm), but we need to handle concrete limitations like `gas limit`.

E.g. for EVM, we have gas limit in EVM context.
So the precompiles should take note of what is the gas limit and how much gas so far has been consumed.
We'd have a struct like:

```Rust
pub EvmExecutionInfo {
  gas_limit: U256,
  used_gas: U256,
  // and more data that is needed
}
```

There would also be a root XVM execution info representation like:
```Rust
pub struct XvmExecutionInfo {
    gas_limit: U256,
    used_gas: U256,
    // and more data that is needed
}
```

And we'd need to convert from one type to another:
```Rust
let evm_exec_info = ....;

let xvm_exec_info: XvmExecutionInfo = evm_exec_info.into();

let parity_wasm_exec_info: ParityWasmExecutionInfo = xvm_exec_info.into();
```
Such conversions should ensure that we keep track of the overall gas limit and how much was used during the calls. As I noted [here](https://github.com/AstarNetwork/astar-frame/pull/81/files#diff-31e0ebbf868cdd39dfea4c054957b66ff5f238240906269a1695a714c8c9b776R95), we need to pass this information around and keep it consistent.

Please correct me if my understanding is incorrect.

Reply:
VM adapter should care about it, not precompile, not xvm don’t known how VM deal with context and parameters.

Reply:
If context wrong - pallet will return error to client, no need to double check it - first on precompile and then again on adapter.

Reply:
Honestly, I’m not sure that EVM precompile can correctly parse/construct all possible XVM contexts without double functionality with XVM adapters.
But yes, we should somehow return consumed gas information to EVM, anyway precompile could get weight from EVM call.